src/main.rs

```diff
@@ -42,68 +63,56 @@ static GLOBAL_BASE: &str = "/home/sam/moods";
-    match flash {
-        Some(m) => Template::render(
-            "index",
-            IndexCtx {
-                dirs: dirs_vec,
-                flash: Some(m.msg().to_string()),
-                flash_name: Some(m.name().to_string()),
-            },
-        ),
-        None => Template::render(
-            "index",
-            IndexCtx {
-                dirs: dirs_vec,
-                flash: None,
-                flash_name: None,
-            },
-        ),
-    }
+fn index(audio_mx: State<AudioMX>, flash_message: Option<FlashMessage>) -> Template {
+    let mp = audio_mx.lock().unwrap();
+    let dirs = mp.keys().map(Clone::clone).collect();
+    let flash = flash_message.as_ref().map(|m| m.msg().to_string());
+    let flash_name = flash_message.as_ref().map(|m| m.name().to_string());
+    Template::render(
+        "index",
+        IndexCtx {
+            dirs,
+            flash,
+            flash_name,
+        },
+    )
 }
+
 #[get("/<subdir>")]
 fn categorize_genre(
     subdir: String,
     audio_mx: State<AudioMX>,
     flash: Option<FlashMessage>,
 ) -> Result<Template, Flash<Redirect>> {
-    let mut context = HashMap::<String, String>::new();
-    let mut map = audio_mx.mp.lock().expect("locking");
-    if let Some(msg) = flash {
-        context.insert("flash_name".to_string(), msg.name().to_string());
-        context.insert("flash".to_string(), msg.msg().to_string());
-    }
-    let lister = match map.get_mut(&subdir) {
-        None => return Err(Flash::error(Redirect::to(uri![index]), "subdir not found")),
-        Some(x) => x,
-    };
-    match lister.clean() {
-        Err(e) => {
-            println!("{}", e.to_string());
-            return Err(Flash::error(Redirect::to(uri![index]), e.to_string()));
-        }
-        Ok(_) => (),
-    };
-    let (filename, sec) = match lister.get_file() {
-        Err(e) if e.kind() == io::ErrorKind::NotFound => {
+    let mut map = audio_mx.lock().unwrap();
+
+    let lister = map
+        .get_mut(&subdir)
+        .ok_or_else(|| Flash::error(Redirect::to(uri![index]), "subdir not found"))?;
+
+    lister.clean();
+
+    let (filename, sec) = lister.get_file().map_err(|e| {
+        if e.kind() == io::ErrorKind::NotFound {
             map.remove(&subdir);
-            return Err(Flash::error(
-                Redirect::to(uri![index]),
-                "No more files in directory",
-            ));
-        }
-        Err(e) => {
+            Flash::error(Redirect::to(uri![index]), "No more files in directory")
+        } else {
             println!("{}", e.to_string());
-            return Err(Flash::error(Redirect::to(uri![index]), e.to_string()));
+            Flash::error(Redirect::to(uri![index]), e.to_string())
         }
-        Ok(x) => x,
+    })?;
+
+    let mut context = hashmap! {
+        "subdir" => subdir,
+        "audio_path" => filename,
+        "audio_sec" => sec,
     };
```

Comment:
Avoid constructing the `HashMap` until we're sure there aren't errors. Also, we can use `&'static str` keys here.